### ISSUE TITLE
Diff highlighting overflows onto the next line (2 chars too long)

### ISSUE THREAD

**Issue by gibfahn** [dandavison/delta#115]
It looks like delta's syntax highlighting is now overrunning the line by two characters.

[![asciicast](https://asciinema.org/a/307651.svg)](https://asciinema.org/a/307651)

**Comment by gibfahn** [comment]
`git show | delta --width=$(($COLUMNS-2))` works, so I'm guessing that what happens is that the diff is indented by 2 chars, and those 2 chars aren't being removed from the line length.

`git bisect` shows the commit that introduced the issue is 185dbd5e87e35a625ceaaea7fd36a35e948e44b1 (unsurprisingly, as that's the one that changed how the text is printed).

**Comment by dandavison** [comment]
Hey @gibfahn, thanks! I'm not seeing this but there's clearly a bug. What are your values of `PAGER` and `BAT_PAGER`? Problems like this in the past have been related to `less` options:

https://github.com/dandavison/delta/issues/61
https://github.com/dandavison/delta/issues/41
https://github.com/dandavison/delta/issues/10

**Comment by dandavison** [comment]
This makes me think, perhaps we should add a `delta --environment` option which prints out all the relevant environment variables and anything else delta is responding to, to help diagnose issues.

**Comment by gibfahn** [comment]
>What are your values of PAGER and BAT_PAGER and LESS? Problems like this in the past have been related to less options:

Yeah I have some stuff :)

```shell
PAGER=less
LESS=--tabs=4 --ignore-case --RAW-CONTROL-CHARS --quit-if-one-screen --status-column --LONG-PROMPT --HILITE-UNREAD --window=-4
BAT_PAGER=less --tabs=4 --ignore-case --RAW-CONTROL-CHARS --quit-if-one-screen --status-column --LONG-PROMPT --HILITE-UNREAD --window=-4
```

Using `git show | LESS= BAT_PAGER= delta` does "fix" things.

<img width="1680" alt="image" src="https://user-images.githubusercontent.com/15943089/76007967-cbb06900-5f06-11ea-9781-78572f642602.png">


<img width="1680" alt="image" src="https://user-images.githubusercontent.com/15943089/76007939-c2270100-5f06-11ea-846e-746f889f43c2.png">



>This makes me think, perhaps we should add a delta --environment option which prints out all the relevant environment variables and anything else delta is responding to, to help diagnose issues.

Good quick way to do it would be something in the issue template that notes the env vars that are often useful. Maybe just `env | grep -E '(PAGER|LESS)'`.

**Comment by Kr1ss-XD** [comment]
@gibfahn FWIW, my best guess is that the `--status-column` option is the culprit.
This indents the whole `less` ouotput by 2 chars to make room for search highlighting and stuff.

**Comment by dandavison** [comment]
Thanks @Kr1ss-XD! I'm happy to try to support custom `less` options. It looks like this is a regression: it was previously fixed at https://github.com/dandavison/delta/pull/43 (and https://github.com/dandavison/delta/pull/22). In retrospect I'm not that surprised as I've recently simplified a lot of stuff related to the terminal color escape codes that are generated. But, I didn't think of it when I merged those PRs and the test suite doesn't have coverage for it (might be tricky).

**Comment by Kr1ss-XD** [comment]
What about something like `LESS+='-+S'` before outputting anything ?  That would effectively disable that option even if the user includes it in their setup.
Not a very elegant way, but maybe you could do something similar ?

**Comment by dandavison** [comment]
TL;DR I should double check the bat code to see if it has been updated since I vendored it and if `bat` supports `--status-column` then I think `delta` should.  

I've been making it a design goal of `delta` that it should behave similarly to `bat`, wherever that makes sense. So we already do that for `BAT_PAGER`, `BAT_THEME`, local syntax and theme binaries, and various CLI options: `--paging`, `--list-themes`, etc.

So ideally I'd like `delta` to support custom `less` options to the same extent as `bat` does. Indeed, the code that `delta` uses to spawn the `less` subprocess is taken from `bat` (and when the `bat` crate starts supplying public library functions, as it intends to: https://github.com/sharkdp/bat/issues/423, then I plan for `delta` to use `bat` as a library for a few things, hopefully including for spawning the pager).

**Comment by Kr1ss-XD** [comment]
Bat appears to have the same glitch.

![2020-03-06-214746_1920x1080_scrot](https://user-images.githubusercontent.com/28707521/76121525-1e1b8380-5ff4-11ea-8624-5156acc7474c.png)

The left pane shows a snippet opened with bat without `--status-column`, on the right hand the same snippet but with this option.

**Comment by dandavison** [comment]
Awesome, thanks for investigating @Kr1ss-XD. Would the `--terminal-width=-2`  type solution in https://github.com/sharkdp/bat/issues/376 work for us here?

**Comment by Kr1ss-XD** [comment]
Well it seems to work for me, yes. :smiley:

**Comment by gibfahn** [comment]
I know it's not a perfect solution, but would just decreasing the width by two if `$LESS` has a `-J` or a `--status-column` option be an option?

I can always just work around this by not setting the option, it's not exactly a huge loss...

**Comment by dandavison** [comment]
> would just decreasing the width by two if $LESS has a -J or a --status-column option be an option?

Right, that's definitely a possibility. But then `bat` output would still be incorrect. Also, it seems that `-N`, `-w`, and `-W` all influence the number of columns that `less` makes available for content.

We have 3 relevant env vars: `PAGER`, `BAT_PAGER`, and `LESS` (any others?). So `delta` would have to contain code inspecting the values of all these env vars, reconciling them and computing a final offset. `bat` also has a `--pager` option, but `-delta` does not yet have this option. It can totally be added if it helps. 

> I can always just work around this by not setting the option, it's not exactly a huge loss...

I don't want people to have to change global behavior of `less` just for delta (i.e. I don't want you to have to change the `LESS` or `PAGER` env vars).

As a quick workaround, could you specify your `less` command line options in `PAGER`, take then out of `LESS`, and set `BAT_PAGER` to the same options but without `--status-column`?

Even if that works, I don't like that it involves duplicating potentially complex sets of command line options. I'm inclined to think that `bat`'s `--terminal-width` solution is good, and I do also like it that people would then not have to take different measures to get this to work for `bat` and `delta`.

**Comment by Kr1ss-XD** [comment]
> We have 3 relevant env vars: PAGER, BAT_PAGER, and LESS (any others?)

A user can set default command line options in their bat config (usually `~/.config/bat/config`; not sure though if that's relevant for delta.

**Comment by dandavison** [comment]
>  I'm inclined to think that bat's --terminal-width solution is good

I should add, `delta` had previously taken the pragmatic step of just always leaving one (or was it two) extra columns in case the pager uses them. It remains to be ascertained why that regressed, and it should probably just be put back the way it was!

**Comment by dandavison** [comment]
> A user can set default command line options in their bat config (usually ~/.config/bat/config); not sure though if that's relevant for delta.

Yes, good point. My feeling so far is that adding that for `delta` would be confusing since many people are going to hard-code their `delta` command line in `~/.gitconfig` (or mercurial, or subversion). However, `delta` can be used perfectly well on `diff -u` output, which would be a use case for its own config file.

**Comment by dandavison** [comment]
@gibfahn @Kr1ss-XD I believe that all problems related to overflowing background colours are fixed in master due to https://github.com/dandavison/delta/pull/165. I had been doing something naive: I was right-padding with space characters, but I believe that the correct approach is to use the ANSI "Erase in Line" escape sequence, which basically instructs a terminal emulator to fill rightwards with the current color. That's what #165 does. I think that it is standard in terminal emulators. The PR does involve Delta automatically setting an env var in the `less` process though: `LESSANSIENDCHARS=mK`, but again, I _think_ that everyone's less is going to support that.

I anyone's able to confirm that master is working well for them that would be great.

**Comment by Kr1ss-XD** [comment]
> I anyone's able to confirm that master is working well for them that would be great.

I confirm. :thumbsup: Well done !

![20200520-155129](https://user-images.githubusercontent.com/28707521/82454510-28590400-9ab2-11ea-9177-c13af654d395.png)

 left side | right side
------------ | ---------------
  0.1.1 | master

**Comment by dandavison** [comment]
Great, thanks @Kr1ss-XD. Closing this; as always, feel free to reopen if any problems remain.

**Comment by gibfahn** [comment]
Yep, can confirm master works like a charm:

<img width="2558" alt="image" src="https://user-images.githubusercontent.com/15943089/82551349-ebdfe380-9b57-11ea-9a90-ab33bea3ab6e.png">
